### PR TITLE
LSP: Provide hover information in Micronaut yaml files.

### PR DESCRIPTION
**Pull request by dbalek**
- `Hover` and `HoverProvider` introduced that allow to resolve a hover information at a given document offset and return its content vial LSP
- Implementations for Java and Micronaut config Yaml files provided

### CODE REVIEW

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/ElementJavadoc.java [hunk 1]
```diff
@@ -371,26 +371,29 @@ private ElementJavadoc(CompilationInfo compilationInfo, Element element, final U
...
+                    c.toPhase(Phase.RESOLVED);
+                    final Element el = handle.resolve(c);
+                    CharSequence doc = getElementDoc(el, c, header, url, false);
+                    if (doc == null) {
+                        computeDocURL(Collections.emptyList(), false, cancel);
+                        doc = header.append(noJavadocFound());
+                    }
+                    return doc.toString();
+                } catch (Exception ex) {
+                    return null;
```
Comment (sdedic) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/ElementJavadoc.java:394]: Maybe log the exception ?

**Review by sdedic — APPROVED** [review]

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/ElementJavadoc.java (hunk 1 above)
Comment (dbalek) [reply java/java.sourceui/src/org/netbeans/api/java/source/ui/ElementJavadoc.java:395]: Fixed.

File: ide/api.lsp/apichanges.xml [hunk 2]
```diff
@@ -47,6 +47,21 @@
 <!-- ACTUAL CHANGES BEGIN HERE: -->
 
 <changes>
+    <change id="HoverProvider">
```
Comment (JaroslavTulach) [inline ide/api.lsp/apichanges.xml:50]: I don't see any `api.lsp` Javadoc at https://bits.netbeans.org/dev/javadoc/

File: ide/api.lsp/apichanges.xml [hunk 3]
```diff
@@ -47,6 +47,21 @@
...
+        <compatibility binary="compatible" source="compatible" addition="yes" deletion="no"/>
+        <description>
+            A <code>Hover</code> class and <code>HoverProvider</code> interface introduced
+            that allows to resolve a hover information at a given document offset and
+            return its content.
+        </description>
+        <class package="org.netbeans.api.lsp" name="Hover"/>
+        <class package="org.netbeans.spi.lsp" name="HoverProvider"/>
+    </change>
     <change id="HyperlinkTypeDefLocationProvider">
```
Comment (JaroslavTulach) [inline ide/api.lsp/apichanges.xml:65]: Is there an `arch.xml` document? It has to be and it should define one `<api type="export" group="stable"/>` to make the Javadoc mention proper stability on its background.

File: ide/api.lsp/src/org/netbeans/spi/lsp/HoverProvider.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import javax.swing.text.Document;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.spi.editor.mimelookup.MimeLocation;
+
+/**
+ * Interface for resolving hover information at a given document position.
+ * Implementations of the interface should be registered in MimeLookup.
+ * <pre>
+ *
+ *  {@code @MimeRegistration(mimeType = "text/foo", service = HoverProvider.class)
```
Comment (JaroslavTulach) [inline ide/api.lsp/src/org/netbeans/spi/lsp/HoverProvider.java:31]: Can you please use `{@codesnippet}`? Also update other examples to use the tag? See [codesnippet4javadoc](https://github.com/jtulach/codesnippet4javadoc).

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
* the `api.lsp` javadoc isn't listed among _stable_ javadocs
* there are no tests for the new functionality
* please use `{@codesnippet}` for all code snippets in this API

File: ide/api.lsp/apichanges.xml (hunk 2 above)
Comment (dbalek) [reply ide/api.lsp/apichanges.xml:50]: Javadoc added.

File: ide/api.lsp/apichanges.xml (hunk 3 above)
Comment (dbalek) [reply ide/api.lsp/apichanges.xml:65]: `arch.xml` document added.

File: ide/api.lsp/src/org/netbeans/spi/lsp/HoverProvider.java (hunk 4 above)
Comment (dbalek) [reply ide/api.lsp/src/org/netbeans/spi/lsp/HoverProvider.java:31]: `{@codesnippet}` used in `Hover.getContent`,  `HyperlinkLocation.resolve`, and `HyperlinkLocation.resolveTypeDefinition` documentation.

Comment (JaroslavTulach) [reply ide/api.lsp/src/org/netbeans/spi/lsp/HoverProvider.java:31]: Why it is not used in `HoverProvider`?

**Review by JaroslavTulach — APPROVED** [review]
I see a test and I trust `netbeans$ ant javadoc` has been verified to produce sane documentation.

**Pull request merged**